src/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.cs

```diff
@@ -10,6 +10,75 @@ namespace System.IO
+        ///
+        /// -or-
+        ///
+        /// <paramref name="mode" />, <paramref name="rights" />, <paramref name="share" /> or <paramref name="options" /> contain an invalid value.
+        /// </exception>
+        /// <remarks><format type="text/markdown"><![CDATA[
+        /// ## Remarks
+        ///
+        /// This method can also throw any of the exceptions that <xref:System.IO.FileStream> can throw.
+        /// ]]></format></remarks>
+        public static FileStream Create(this FileInfo fileInfo, FileMode mode, FileSystemRights rights, FileShare share, int bufferSize, FileOptions options, FileSecurity fileSecurity)
+        {
+            string badArg = null;
+
+            if (mode < FileMode.CreateNew || mode > FileMode.Append)
+                badArg = nameof(mode);
+
+            if (rights < FileSystemRights.ReadData || rights > FileSystemRights.FullControl)
+                badArg = nameof(rights);
+
+            if (share < FileShare.None || share > FileShare.Inheritable)
+                badArg = nameof(share);
+
+            if (options < FileOptions.None || options > FileOptions.Asynchronous)
+                badArg = nameof(options);
+
+            if (badArg != null)
+                throw new ArgumentOutOfRangeException(badArg, SR.ArgumentOutOfRange_Enum);
+
+            if (bufferSize <= 0)
+                throw new ArgumentOutOfRangeException(nameof(bufferSize));
+
+            if (fileSecurity == null)
+                throw new ArgumentNullException(nameof(fileSecurity));
+
+            // Ignore these two calls. They were my experiments to understand why we could not resolve the issue with what we currently have
+            // fileInfo.SetAccessControl(fileSecurity);
+            // return new FileStream(fileInfo.FullName, mode, mode == FileMode.Append ? FileAccess.Write : FileAccess.ReadWrite, share, bufferSize, options);
+
+            return null;
+        }
+
+        /// <summary>Creates a new directory, ensuring it is created with the specified directory security.</summary>
+        /// <param name="directoryInfo">The object describing a directory that does not exist in disk yet.</param>
+        /// <param name="directorySecurity">An object that determines the access control and audit security for the directory.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="directorySecurity" /> is <see langword="null" />.</exception>
+        public static void Create(this DirectoryInfo directoryInfo, DirectorySecurity directorySecurity)
+        {
+            if (directorySecurity == null)
+                throw new ArgumentNullException(nameof(directorySecurity));
+
+            // I cannot do this (which is what FileSystem.CreateDirectory() currently does)
+            // unless we want a circular dependency between System.IO.FileSystem and System.IO.FileSystem.AccessControl
+
```

Comment:
@JeremyKuhne this is what's causing me conflict: I am already consuming System.IO.FileSystem.AccessControl as a new dependency for System.IO.FileSystem, so I cannot call FileSystem.CreateDirectory here unless I have a circular dependency. The other problem I have is that I cannot invalidate the directoryInfo here because Invalidate() is private.

Reply:
Update: I think I will have to do something like in [this PR](https://github.com/dotnet/corefx/pull/26395/files), where the `System.IO.Pipes.AccessControl` APIs are declared in the ref files, but the source code is written in `System.IO.Pipes`.

I noticed this could be the solution after I submitted [my other PR](https://github.com/dotnet/corefx/pull/41838) for the implementation of the creation APIs for Pipes.

Reply:
Just include the sources. I gave a sample on how you should change the signature.

Reply:
Ok I think I have covered this part, all the sources were included (including the dependencies of these dependencies). I only need to figure out one error, and I'll push a new commit.